codalab/client/json_api_client.py

```diff
@@ -0,0 +1,299 @@
+                'attributes': {
+                    'name': 'hello'
+                },
+                'relationships': {
+                    'owner': {
+                        'data': {
+                            'id': '345',
+                            'type': 'users'
+                        }
+                    },
+                    'parent': {
+                        'data': {
+                            'id': '567',
+                            'type': 'bundles'
+                        }
+                    }
+                },
+            },
+            'included': [
+                {
+                    'type': 'users':
+                    'id': '345',
+                    'attributes': {
+                        'name': 'percy',
+                        'affiliation': 'stanford'
+                    }
+                }
+            ]
+        })
+
+     => {
+            'owner': {
+                'id': '345',
+                'name': 'percy',
+                'affiliation: 'stanford'
+            },
+            'parent': {
+                'id': '567'
+            },
+            'id': '123',
+            'name': 'hello'
+        }
+
+    :param document: the JSON-API payload as a dict
+    :return:
+    """
+    def unpack_object(obj_data):
+        obj = obj_data['attributes']
+        obj['id'] = obj_data['id']
+
+        if 'meta' in obj_data:
+            obj['meta'] = obj_data['meta']
+
+        for key, relationship in obj_data.get('relationships', {}).iteritems():
+            if isinstance(relationship['data'], list):
+                obj[key] = []
+                for linkage in relationship['data']:
+                    if linkage['id'] in included[linkage['type']].keys():
+                        obj[key].append(unpack_object(included[linkage['type']][linkage['id']]))
+                    else:
+                        obj[key].append({'id': linkage['id']})
+            else:
+                linkage = relationship['data']
+                if linkage is None:
+                    obj[key] = None
+                elif linkage['id'] in included[linkage['type']]:
+                    obj[key] = unpack_object(included[linkage['type']][linkage['id']])
+                else:
+                    obj[key] = {'id': linkage['id']}
+
+        return obj
+
+    included = defaultdict(dict)
+    for resource in document.get('included', []):
+        included[resource['type']][resource['id']] = resource
+
+    data = document.get('data', None)
+    if isinstance(data, list):
+        result = [unpack_object(data) for data in document['data']]
+    elif isinstance(data, dict):
+        result = unpack_object(document['data'])
+    else:
+        result = {}
+
+    # Include meta
+    # (Warning: this may overwrite meta present at the resource object level.)
+    if 'meta' in document:
+        result['meta'] = document['meta']
+
+    return result
+
+
+def pack_document(objects, type_):
```

Comment:
This packing and unpacking of documents seems like premature optimization... Just my two cents.